atomix/cluster/src/test/java/io/atomix/raft/RaftRule.java

```diff
@@ -0,0 +1,521 @@
+import io.atomix.raft.zeebe.ZeebeLogAppender;
+import io.atomix.storage.StorageLevel;
+import io.atomix.storage.journal.Indexed;
+import io.atomix.storage.journal.JournalReader.Mode;
+import io.atomix.utils.concurrent.SingleThreadContext;
+import io.atomix.utils.concurrent.ThreadContext;
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BooleanSupplier;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.junit.rules.ExternalResource;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public final class RaftRule extends ExternalResource {
+
+  public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+  private volatile int nextId;
+  private volatile List<RaftMember> members;
+  private Map<String, Long> memberLog;
+  private final List<RaftServer> servers = new ArrayList<>();
+  private volatile TestRaftProtocolFactory protocolFactory;
+  private volatile ThreadContext context;
+  private Path directory;
+  private final int nodeCount;
+  private volatile long highestCommit;
+  private final AtomicReference<CommitAwaiter> commitAwaiterRef = new AtomicReference<>();
+  private long position;
+
+  private RaftRule(final int nodeCount) {
+    this.nodeCount = nodeCount;
+  }
+
+  public static RaftRule withBootstrappedNodes(final int nodeCount) {
+    if (nodeCount < 1) {
+      throw new IllegalArgumentException("Expected to have at least one node to configure.");
+    }
+    return new RaftRule(nodeCount);
+  }
+
+  public static RaftRule withoutNodes() {
+    return new RaftRule(-1);
+  }
+
+  @Override
+  public Statement apply(final Statement base, final Description description) {
+    final var statement = super.apply(base, description);
+    return temporaryFolder.apply(statement, description);
+  }
+
+  @Override
+  protected void before() throws Throwable {
+    directory = temporaryFolder.newFolder().toPath();
+
+    position = 0;
+    members = new ArrayList<>();
+    memberLog = new ConcurrentHashMap<>();
+    nextId = 0;
+    context = new SingleThreadContext("raft-test-messaging-%d");
+    protocolFactory = new TestRaftProtocolFactory(context);
+
+    if (nodeCount > 0) {
+      createServers(nodeCount);
+    }
+  }
+
+  @Override
+  protected void after() {
+    servers.forEach(
+        s -> {
+          try {
+            if (s.isRunning()) {
+              s.shutdown().get(30, TimeUnit.SECONDS);
```

Comment:
Not sure why there is a timeout. The exception will be ignored anyway. And if this runs sequentially, then each server could take up to 30 seconds. Maybe better collect the futures and wait for all of them. Or just call shutdown without waiting